com.avaloq.tools.ddk.xtext.generator/src/com/avaloq/tools/ddk/xtext/generator/parser/antlr/ParserSemanticPredicatesUtil.java

```diff
@@ -0,0 +1,628 @@
+      if (groupElement instanceof Action) {
+        continue;
+      }
+      return Objects.equal(groupElement, element); // NOPMD AvoidBranchingStatementAsLastInLoop
+    }
+    return false;
+  }
+
+  /**
+   * Keyword rule proposals.
+   *
+   * @param rule
+   *          Xtext grammar rule
+   * @return A string containing the proposals for abstract content assist
+   */
+  public static String keywordRuleProposals(final AbstractRule rule) {
+    String text = getText(rule);
+    if (text != null) {
+      Matcher matcher = KEYWORD_RULE_ANNOTATION_PATTERN.matcher(text);
+      if (matcher.find()) {
+        StringBuilder predicate = new StringBuilder();
+        boolean passedFirst = false;
+        for (String kw : KEYWORDS_SPLITTER.split(matcher.group(1))) {
+          if (passedFirst) {
+            predicate.append("    "); //$NON-NLS-1$
+          }
+          predicate.append("propose("); //$NON-NLS-1$
+          predicate.append('"');
+          predicate.append(kw);
+          predicate.append('"');
+          predicate.append(", "); //$NON-NLS-1$
+          predicate.append('"');
+          predicate.append(kw);
+          predicate.append('"');
+          predicate.append(", getImage(model), context, acceptor);"); //$NON-NLS-1$
+          predicate.append('\n');
+          passedFirst = true;
+        }
+        if (passedFirst) {
+          return predicate.toString();
+        }
+      }
+    }
+    return ""; //$NON-NLS-1$
+  }
+
+  /**
+   * Checks if the given rule contains a keyword rule annotation.
+   *
+   * @param rule
+   *          Grammar rule
+   * @return {@code true} if there is an annotation, {@code false} otherwise
+   */
+  public static boolean isKeywordRulePredicate(final AbstractRule rule) {
+    String text = getText(rule);
+    if (text != null) {
+      Matcher matcher = KEYWORD_RULE_ANNOTATION_PATTERN.matcher(text);
+      if (matcher.find()) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Checks if the given rule contains a keyword rule annotation.
+   *
+   * @param rule
+   *          Grammar rule
+   * @return {@code true} if there is an annotation, {@code false} otherwise
+   */
+  public static boolean isSemanticPredicate(final AbstractRule rule) {
+    String text = getText(rule);
+    if (text != null) {
+      Matcher matcher = SEMANTIC_PREDICATE_PATTERN.matcher(text);
+      if (matcher.find()) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Checks if the given rule contains a NoBacktrack annotation.
+   *
+   * @param rule
+   *          Grammar rule
+   * @return {@code true} if there is an annotation, {@code false} otherwise
+   */
+  public static boolean isNoBacktrack(final AbstractRule rule) {
+    String text = getText(rule);
+    if (text != null) {
+      Matcher matcher = NO_BACKTRACK_ANNOTATION_PATTERN.matcher(text);
+      if (matcher.find()) {
+        return true;
+      }
+    }
+    return false;
+  }

```

Comment:
These functions are identical if it not where for the used pattern, could we create a private method, for example

  private static boolean matches(final AbstractRule rule, final Pattern pattern) {
    String text = getText(rule);
    if (text != null) {
      Matcher matcher = pattern.matcher(text);
      if (matcher.find()) {
        return true;
      }
    }
    return false;    
  }

  /**
   * Checks if the given rule contains a keyword rule annotation.
   *
   * @param rule
   *          Grammar rule
   * @return {@code true} if there is an annotation, {@code false} otherwise
   */
  public static boolean isKeywordRulePredicate(final AbstractRule rule) {
    return matches(rule, KEYWORD_RULE_ANNOTATION_PATTERN);
  }

  /**
   * Checks if the given rule contains a keyword rule annotation.
   *
   * @param rule
   *          Grammar rule
   * @return {@code true} if there is an annotation, {@code false} otherwise
   */
  public static boolean isSemanticPredicate(final AbstractRule rule) {
    return matches(rule, SEMANTIC_PREDICATE_PATTERN);
  }